### PR TITLE
[LI-HOTFIX] Improve the efficiency of replica deletion

### PR DESCRIPTION
**Pull request by gitlw**
TICKET = KAFKA-13815
LI_DESCRIPTION =
KAFKA-10002 identified that deletion of replicas can be slow when a StopReplica request is being
processed, and has implemented a change to improve the efficiency.
We found that the efficiency can be further improved by avoiding the reinitialization of the
leader epoch cache and partition metadata for a replica that needs to be deleted.

EXIT_CRITERIA = When KAFKA-13815 is resolved and the change is pulled in

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/log/Log.scala [hunk 1]
```diff
@@ -755,8 +755,10 @@ class Log(@volatile private var _dir: File,
           producerStateManager.updateParentDir(dir)
           // re-initialize leader epoch cache so that LeaderEpochCheckpointFile.checkpoint can correctly reference
           // the checkpoint file in renamed log directory
-          initializeLeaderEpochCache()
-          initializePartitionMetadata()
+          if (reinitialize) {
```
Comment (CCisGG) [inline core/src/main/scala/kafka/log/Log.scala]: nit: "shouldReinitialize" might be better name for a boolean var.

**Comment by CCisGG** [comment]
> EXIT_CRITERIA = When [KAFKA-13815](https://issues.apache.org/jira/browse/KAFKA-13815) is resolved and the change is pulled in

@gitlw I thought the exit_criteria means when is this commit not needed anymore. For this PR though, it exactly solve the issue of kafka-13815, which shouldn't be defined as an exit_criteria. Did I understand it wrong?

File: core/src/main/scala/kafka/log/Log.scala (hunk 1 above)
Comment (gitlw) [reply core/src/main/scala/kafka/log/Log.scala]: Thanks for the suggestion. I've renamed the variable.

**Comment by gitlw** [comment]
> @gitlw I thought the exit_criteria means when is this commit not needed anymore. For this PR though, it exactly solve the issue of [kafka-13815](https://issues.apache.org/jira/browse/KAFKA-13815), which shouldn't be defined as an exit_criteria. Did I understand it wrong?

I think the EXIT_CRITERIA field is still pretty useful and should be kept. 
EXIT_CRITERIA means when we can discard the change from the hotfix list, e.g. when the change no longer applies, or when the change is available from upstream kafka (at which point it doesn't need to be tracked as a hotfix).

**Review by CCisGG — APPROVED** [review]

**Comment by lmr3796** [comment]
> > EXIT_CRITERIA = When [KAFKA-13815](https://issues.apache.org/jira/browse/KAFKA-13815) is resolved and the change is pulled in
> 
> @gitlw I thought the exit_criteria means when is this commit not needed anymore. For this PR though, it exactly solve the issue of [kafka-13815](https://issues.apache.org/jira/browse/KAFKA-13815), which shouldn't be defined as an exit_criteria. Did I understand it wrong?

@CCisGG For the ticket, the patch must be an upstream patch to resolve it.  So EXIT_CRITERIA is actually correctly used this way.

**Review by lmr3796 — APPROVED** [review]
LGTM

**Comment by lmr3796** [comment]
This commit was later reverted

**Pull request merged**